### DISCUSSION TITLE
Printing select statement with bound parameter values

### DISCUSSION THREAD

**Discussion by molotch** [sqlalchemy/sqlalchemy#8792] [Usage Questions] [answered]
Hi,

I'm trying to print my select statements with parameter values bound.

I have two code paths with different values that shares 90% of the code. Path 1 works fine, I get the select statement in text with bound parameter values. In Path 2 I get the error 'tuple' object has no attribute 'replace' and the stack trace is only pointing at the print line.

> **Exception has occurred: AttributeError** X
> 'tuple' object has no attribute 'replace'

```python
print(stmt.cte().compile(dialect=postgresql.dialect(), compile_kwargs={"literal_binds": True}))
```

If I print the statement without binding parameters (literal_binds) it works fine.

How do I find out which value it tries to bind to which parameter? I looked into the statement object, but that had way to many attributes to be helpful. I've tried looking through all the parameter values and nothing looks out of the ordinary.

This is also a suggestion for improvement of the error messages I reckon.

**Comment by zzzeek** [comment]
hi - 

I have no idea what the problem is, well, it's probably something that's in the query you've built.   But no way to know at least with the stack trace.

**Reply by zzzeek** [reply]
and of course, some suggestion of what the statement looks like.

**Reply by molotch** [reply]
Thank you for replying. These are the last lines of the stacktrace.

>File "c:\Users\...\.venv\lib\site-packages\sqlalchemy\sql\compiler.py", line 2564, in render_literal_bindparam
>    return self.render_literal_value(value, bindparam.type)
>File "c:\Users\...\.venv\lib\site-packages\sqlalchemy\dialects\postgresql\base.py", line 2379, in render_literal_value
>    value = super(PGCompiler, self).render_literal_value(value, type_)
>File "c:\Users\...\.venv\lib\site-packages\sqlalchemy\sql\compiler.py", line 2579, in render_literal_value
>    return processor(value)
>File "c:\Users\...\.venv\lib\site-packages\sqlalchemy\sql\sqltypes.py", line 310, in process
>    value = value.replace("'", "''")
>AttributeError: 'tuple' object has no attribute 'replace'

The statement is a select with a few joins and filters. So some join conditions and some of the filters are literals. I've tried to check them all and can't find anything particularly strange.

I had a problem with date not being able to be bound without converting it to string beforehand. But then I got a error message that actually made sense. Now I believe I only have string and int values for all literals.

Maybe I can install sqlalchemy in edit mode and set a breakpoint in sqltypes.py on line 310 and see on which literal it throws?

**Reply by zzzeek** [reply]
you have a tuple assigned as a parameter somewhere in your statement inappropriately.  look for something like ``column == x,`` or similar.

**Reply by molotch** [reply]
No luck, went through all filters and joins, there's no literals that's not wrapped by the literal() function. There's no property in the statement object I can inspect if I set a breakpoint? Somewhere all the parameters and their values must be stored?

**Reply by zzzeek** [reply]
sure, go into the source and set breakpoints.   

im not sure if you realize, I cannot help you if you can't show me actual code that shows the problem, I have no idea what you are doing.

**Reply by molotch** [reply]
I realize that, but there's quite alot of code that builds the "query" so I wouldn't know what to copy paste. If you have any suggestion on what to look for besides `column == x`  I could maybe try to copy paste something. What I don't understand is how column comparisons can become a tuple instead of an "expression", most of them work, it's apparently just one somewhere that doesn't work.

**Reply by molotch** [reply]
So I found it, finally. We use Enum values to represent the filter values. Someone had put a , on one of the enum values which Black formatted into ("value",) and when that particular value was chosen anywhere in the pipeline the tuple error occured. Thank you for your help.

**Reply by molotch** [reply]
I suggest you should emit which parameter and value the bind fails on when this error occurs. Would help immensely in troubleshooting.

**Reply by CaselIT** [reply]
> I suggest you should emit which parameter and value the bind fails on when this error occurs. Would help immensely in troubleshooting.

it's not that easy to add it in the sqltype, but maybe the exception could be catched an re-raised with more information by the compiler.

**Reply by zzzeek** [reply]
I've added #8800 where we have a statement like this:

```py
from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)


stmt = select(A).where(A.data.in_([('adf', )]) )

print(stmt.compile(compile_kwargs={"literal_binds": True}))
```

will raise:

```
sqlalchemy.exc.CompileError: Could not render literal value "('adf',)" for bound parameter %(140259383958528 data)s with datatype VARCHAR; see parent stack trace for more detail.
```

**Reply by zzzeek** [reply]
the exception message explicitly refers to "parent stack trace" which those vanishingly few py2 users left won't actually see.

**Reply by CaselIT** [reply]
if it's v2 only the py2 users are no more :)

**Comment by molotch** [comment, marked as answer]
We use Enum values to represent the filter values. Someone had put a , behind one of the enum values which Black formatted into ("value",) and when that particular value was chosen anywhere in the pipeline the tuple error occured.
